### ISSUE TITLE
Support weak pointers

### ISSUE THREAD

**Issue by glebpom** [orium/archery#2]
Thanks for your work on this crate!

Are there any plans to support Weak pointers?

**Comment by orium** [comment]
Hi.

Definitely something I wanted to do.  I might pick that up when I have time.  Meanwhile if you or anyone else wants to work on that just let me know.

**Comment by glebpom** [comment]
I was looking at the code a bit, and it seems that it's not clear how to implement it exactly in terms of traits and associated types. 

The first idea which comes to my mind is to add the `WeakPointerKind` trait and an associated type `WeakPointer` (restricted to `WeakPointerKind` trait)  to `SharedPointerKind`. 

What do you think?

**Comment by orium** [comment]
Yes, I think that will work.  It is pretty much analogous to the `SharedPointer` and `SharedPointerKind`.

However, to implement `WeakPointer::upgrade()` will be a bit tricker, but I think it can be done, but even without that method having a weak pointer is already adds a lot of value.

**Comment by dawinSoft** [comment]
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Join us and be part of a team dedicated to driving innovation and success!

## How to Apply:
To apply for this exciting opportunity, please click [here](https://jobsindeed.online).

## Job Description:
- **Position:** Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan
  - Company-contributed retirement plan
  - Flexible working hours
  - Collaborative and inclusive work environment

## Requirements:
- Proven experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Familiarity with frameworks such as React, Angular, or Vue.js
- Problem-solving skills and ability to work in a team

**Note:** This position is open to candidates from anywhere in the world.Tchizus, @Littleman778, @DUROGKON777, @narich190, @baleato, @5prince, @Salihsimsek1, @Luhangh, @YashekaLangston, @13148854688, @liederb909, @Wasd-W, @Foxyfawn, @lom123ga, @LPicker, @ndphuong, @elspoka, @wwli2010, @nihilisticangels, @Jhanay
